core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -424,38 +424,46 @@ class KafkaApis(val requestChannel: RequestChannel,
         }
       }
 
-      def produceResponseCallback(bandwidthThrottleTimeMs: Int) {
-        if (produceRequest.acks == 0) {
-          // no operation needed if producer request.required.acks = 0; however, if there is any error in handling
-          // the request, since no response is expected by the producer, the server will close socket server so that
-          // the producer client will know that some error has happened and will refresh its metadata
-          if (errorInResponse) {
-            val exceptionsSummary = mergedResponseStatus.map { case (topicPartition, status) =>
-              topicPartition -> status.error.exceptionName
-            }.mkString(", ")
-            info(
-              s"Closing connection due to error during produce request with correlation id ${request.header.correlationId} " +
-                s"from client id ${request.header.clientId} with ack=0\n" +
-                s"Topic and partition to exceptions: $exceptionsSummary"
-            )
-            closeConnection(request, new ProduceResponse(mergedResponseStatus.asJava).errorCounts)
-          } else {
-            sendNoOpResponseExemptThrottle(request)
-          }
+      // When this callback is triggered, the remote API call has completed
+      request.apiRemoteCompleteTimeNanos = time.nanoseconds
+
+      // Record both bandwidth and request quota-specific values and throttle by muting the channel if any of the quotas
+      // have been violated. If both quotas have been violated, use the max throttle time between the two quotas. Note
+      // that the request quota is not enforced if acks == 0.
+      val bandwidthThrottleTimeMs = quotas.produce.maybeRecordAndGetThrottleTimeMs(request, numBytesAppended)
+      val requestThrottleTimeMs = if (produceRequest.acks == 0) 0 else quotas.request.maybeRecordAndGetThrottleTimeMs(request)
+      val maxThrottleTimeMs = Math.max(bandwidthThrottleTimeMs, requestThrottleTimeMs)
```

Comment:
I am not sure about this. Since quotas are calculated differently, not sure `max` gives the same result as the throttle times calculated separately as it was done earlier.

Reply:
With this KIP, a response needs to be sent out with a throttle time value before actually starting throttling, and thus we need to determine how long we should mute the channel first before sending out the response. In other words, if multiple quotas are violated, we can't really wait till throttling for one quota is over before computing throttle time for next quota.
 
Under this scenario, I thought using the max is reasonable. It will be same as before if only one quota is violated. If multiple quotas are violated, using max may not give the same throttle time
in some cases but I am not sure what will be a better alternative.

I am not entirely sure about how throttle times can be "calculated separately" with this KIP. Are you suggesting that on multiple quota violations, say, produce and request, we throttle for produce violation only and deal with any remaining request quota violation the next time the client sends another produce request after the initial throttling? But this is also different from the way it was and can be inefficient (in case we keep picking smaller throttle time).

Reply:
No, I am not suggesting that we do only one at a time. I think we need to see if we can do a better calculation than `max` to combine the two (or potentially many in future).

Reply:
I still think using max is reasonable. In my understanding, that is the minimum amount of time we need to throttle anyway for the traffic at the time of quota violation. There may be other connections using the same client id/user while the throttling is going on, but that will only add more load and thus will not improve the throttle time. I also discussed with @becketqin and he agreed. Would you let me know if you have any specific suggestions?